### PR TITLE
bzlmod prototype

### PR DESCRIPTION
**Pull request by fmeum**
Modules can use the `go_dep` module extension to define Go dependencies.

Only Go module dependencies are supported for now (`go_dep.mod`). HTTP
and VCS Go repositories can be implemented for the root module once
https://github.com/bazelbuild/bazel/pull/15792 has been merged. They
should follow the semantics of archive_override/git_override.

### CODE REVIEW

**Comment by fmeum** [comment]
@achew22 @linzhp The prototype should work reasonably well now. The PR is still based on #1294.

File: internal/bzlmod/go_deps.bzl [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        # dependency graphs of all transitive Bazel module dependencies, which
+        # is not what `go mod` does. But since this algorithm ends up using only
+        # Go module versions that have been explicitly declared somewhere in the
+        # full graph, we can assume that at that place all its required
+        # transitive dependencies have also been declared - we may end up
+        # resolving them to higher versions, but only compatible ones.
+        importpaths = {}
+        for module_tag in module.tags.module:
+            if module_tag.importpath in importpaths:
+                fail("Duplicate importpath '%s' in module '%s'" % (module_tag.importpath, module.name))
```
Comment (linzhp) [inline internal/bzlmod/go_deps.bzl:64]: Prefer:
```suggestion
                fail("Duplicate importpath '{}' in module '{}'".format(module_tag.importpath, module.name))
```

File: internal/bzlmod/go_deps.bzl [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    candidate_name = "_".join(segments).replace("-", "_")
+    return "".join([c.lower() if c.isalnum() else "_" for c in candidate_name.elems()])
+
+def _go_repository_directives_impl(ctx):
+    directives = [
+        "# gazelle:repository go_repository name={name} {directives}".format(
+            name = name,
+            directives = " ".join(directives),
+        )
+        for name, directives in ctx.attr.directives.items()
```
[the remark this answers is not in the document]
Comment (fmeum) [reply internal/bzlmod/go_deps.bzl]: Fixed.

File: internal/bzlmod/go_deps.bzl (hunk 1 above)
Comment (fmeum) [reply internal/bzlmod/go_deps.bzl]: Thanks, fixed.

File: internal/bzlmod/go_deps.bzl [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        # dependency graphs of all transitive Bazel module dependencies, which
+        # is not what `go mod` does. But since this algorithm ends up using only
+        # Go module versions that have been explicitly declared somewhere in the
+        # full graph, we can assume that at that place all its required
+        # transitive dependencies have also been declared - we may end up
+        # resolving them to higher versions, but only compatible ones.
+        importpaths = {}
+        for module_tag in module.tags.module:
+            if module_tag.importpath in importpaths:
+                fail("Duplicate importpath '{}' in module '{}'".format(module_tag.importpath, module.name))
```
[the remark this answers is not in the document]
Comment (fmeum) [reply internal/bzlmod/go_deps.bzl:64]: Yes, there is nothing stopping a module from doing:
```
go_deps.module(importpath = "example.org/foo", version = "1.0.0")
go_deps.module(importpath = "example.org/foo", version = "1.1.0")
```

We could still apply MVS, but I would prefer emitting an error. Once we get to overrides, we would definitely get ambiguities.

**Pull request merged**